Sources/Prelude/Refinements.swift

```diff
@@ -0,0 +1,207 @@
+///
+/// - Parameter refined: a value that is refined by two rules
+/// - Returns: the same value, refined only by the rule on the right
+public func right<A, L, R>(_ refined: Refined<A, Both<L, R>>) -> Refined<A, R> {
+    guard let result = R.of(refined.value) else {
+        preconditionFailure("when calling `right(_:)`: the value “\(refined.value)” didn’t satisfy the refinement “\(R.self)”. This shouldn’t happen. Was the value mutated after refinements had already been checked?")
+    }
+    return result
+}
+
+// MARK: - Not
+
+/// a refinement that can be used to invert any other refinement
+public enum Not<Rule: Refinement>: Refinement {
+    public typealias RefinedType = Rule.BaseType
+    public static func isValid(_ value: Rule.BaseType) -> Bool {
+        return Rule.predicate.complement.contains(value)
+    }
+}
+
+// MARK: - OneOf
+
+/// a refinement that can be used to combine any two other refinements (`L` and `R`), such that either one or the other must pass
+public enum OneOf<L: Refinement, R: Refinement>: Refinement where L.BaseType == R.BaseType {
+    public typealias RefinedType = L.BaseType
+    public static func isValid(_ value: RefinedType) -> Bool {
+        return L.predicate.union(R.predicate).contains(value)
+    }
+}
+
+/// given a value that is refined by a `OneOf<L, R>` (either the refinement `L`, or the refinement `R`), produce a value that is just refined by `L`. This function returns nil if the underlying value does not satisfy `L`
+///
+/// - Parameter refined: a value that is refined by the disjunction of two rules
+/// - Returns: the same value, refined only by the rule on the left, or nil
+public func left<A, L, R>(_ refined: Refined<A, OneOf<L, R>>) -> Refined<A, L>? {
+    return L.of(refined.value)
+}
+
+/// given a value that is refined by a `OneOf<L, R>` (either the refinement `L`, or the refinement `R`), produce a value that is just refined by `R`. This function returns nil if the underlying value does not satisfy `R`
+///
+/// - Parameter refined: a value that is refined by the disjunction of two rules
+/// - Returns: the same value, refined only by the rule on the right, or nil
+public func right<A, L, R>(_ refined: Refined<A, OneOf<L, R>>) -> Refined<A, R>? {
+    return R.of(refined.value)
+}
+
+/// given a value that is refined by a `OneOf<L, R>` (either the refinement `L`, or the refinement `R`), produce a value that is refined by both (`Both<L, R>`). This function returns nil if the underlying value does not satisfy both `L` and `R`
+///
+/// - Parameter refined: a value that is refined by the disjunction of two rules
+/// - Returns: the same value, refined by both of the rules, or nil
+public func both<A, L, R>(_ refined: Refined<A, OneOf<L, R>>) -> Refined<A, Both<L, R>>? {
+    return Both<L, R>.of(refined.value)
+}
+
+// MARK: - Array
+
+public extension Array {
+    enum NonEmpty: Refinement {
```

Comment:
Could we potentially use this to build in something simialr to pointfree's NonEmpty type? that would be pretty sweet, and I can think of some interesting applications we could pitch to other devs once it's in.

Reply:
Ah, good question. This is — as I understand it anyway — a fundamentally different implementation from theirs. In fact I think this was used as an example in the video (or blog post, whatever) of a non-optimal way to do this.

We totally could do “real” `NonEmpty` but then I think we’d want to remove this. Actually the more I think about it the more I am convinced to take this out and do real NonEmpty later…